url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/grammaticalbehaviors/GEBT_Mario/Graph.java

```diff
@@ -0,0 +1,1289 @@
+        }
+
+    }
+
+
+    public int addNode(int a_x, int a_y)
+    {
+        //calculate an unique ID depending on the coordinates:
+        int id = 100000*(100+a_y) + (10000+a_x);
+        Integer idKey = new Integer(id);
+
+        if(!m_nodes.containsKey(idKey))
+        {
+            //Not registered, create node and insert it.
+            Node newNode = new Node(id, a_x, a_y);
+            m_nodes.put(idKey, newNode);
+
+            postProcessNode(newNode, a_x, a_y);
+            m_nodeCounter++;
+        }
+        
+        return id;
+    }
+
+    //Checks if there is an existing node in the given coordinates.
+    //If it does not exist, returns -1. Otherwise, return its ID.
+    public int existsNode(int a_x, int a_y)
+    {
+    //calculate an unique ID depending on the coordinates:
+        int id = 100000*(100+a_y) + (10000+a_x);
+        Integer idKey = new Integer(id);
+
+        if(!m_nodes.containsKey(idKey))
+            return -1;
+        else return id;
+    }
+
+
+    //Checks if there is an existing node with the given ID.
+    //If it does not exist, returns -1. Otherwise, return its ID.
+    public int existsNode(int a_id)
+    {
+        if(!m_nodes.containsKey(a_id))
+            return -1;
+        else return a_id;
+    }
+
+    public void addBreakableEdge(int a_aID, int a_bID, int a_mode)
+    {
+        Node nodeA = m_nodes.get(a_aID);
+        Node nodeB = m_nodes.get(a_bID);
+
+        if(nodeA != null && nodeB != null)
+        {
+            int aY = nodeA.getY();
+            int aX = nodeA.getX();
+            int bY = nodeB.getY();
+            int bX = nodeB.getX();
+
+            int distance = nodeA.manhattanDistanceTo(nodeB);
+            Edge newEdge = null;
+            Long idKey = new Long(0);
+
+            //We need to see which one is the lower one
+            if(nodeA.getY() < nodeB.getY())
+            {
+                //A under B
+
+                //calculate an unique ID depending on the IDs of the nodes:
+                long id = (100000000L * a_aID) + a_bID;
+                idKey = new Long(id);
+
+                if(!m_edges.containsKey(idKey))
+                {
+                    newEdge = new Edge(id, a_aID, a_bID, a_mode, distance);
+                    nodeA.addEdge(id);
+                    m_edges.put(idKey, newEdge);
+                    m_edgeCounter++;
+                }
+            }
+            else
+            {
+                //B under A
+
+                //calculate an unique ID depending on the IDs of the nodes:
+                long id = (100000000L * a_bID) + a_aID;
+                idKey = new Long(id);
+
+                if(!m_edges.containsKey(idKey))
+                {
+                    newEdge = new Edge(id, a_bID, a_aID, a_mode, distance);
+                    nodeB.addEdge(id);
+                    m_edges.put(idKey, newEdge);
+                    m_edgeCounter++;
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 158:202, 214:258.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.